modules/core/arrow-annotations-processor/src/main/java/arrow/generic/CoproductFileGenerator.kt

```diff
@@ -0,0 +1,194 @@
+                        .returns(parameterizedCoproductNClassName(generics))
+                        .build()
+        )
+    }
+}
+
+private fun FileSpec.Builder.addSelectFunctions(generics: List<String>) {
+    addImport("arrow.core", "Option")
+    addImport("arrow.core", "toOption")
+
+    for (generic in generics) {
+        val receiverGenerics = generics
+                .map { if (it == generic) TypeVariableName(generic) else TypeVariableName("*") }
+                .toTypedArray()
+        val additionalParameterCount = generics.indexOf(generic)
+
+        addFunction(
+                FunSpec.builder("select")
+                        .addAnnotations(additionalParameterSuppressAnnotation(additionalParameterCount))
+                        .addTypeVariable(TypeVariableName(generic))
+                        .receiver(ClassName("", "Coproduct${generics.size}").parameterizedBy(*receiverGenerics))
+                        .addParameters(additionalParameterSpecs(additionalParameterCount))
+                        .returns(ClassName("arrow.core", "Option").parameterizedBy(TypeVariableName(generic)))
+                        .addStatement("return (this as? ${genericsToClassNames[generic]})?.${generic.toLowerCase()}.toOption()")
+                        .build()
+        )
+    }
+}
+
+private fun FileSpec.Builder.addFoldFunction(generics: List<String>) {
+    addFunction(
+            FunSpec.builder("fold")
+                    .receiver(parameterizedCoproductNClassName(generics))
+                    .addTypeVariables(generics.toTypeParameters() + TypeVariableName("RESULT"))
+                    .addParameters(
+                            generics.map {
+                                ParameterSpec.builder(
+                                        it.toLowerCase(),
+                                        LambdaTypeName.get(
+                                                parameters = listOf(ParameterSpec.unnamed(TypeVariableName(it))),
+                                                returnType = TypeVariableName("RESULT")
+                                        )
+                                ).build()
+                            }
+                    )
+                    .returns(TypeVariableName("RESULT"))
+                    .apply {
+                        beginControlFlow("return when (this)")
+                        for (generic in generics) {
+                            addStatement("is ${genericsToClassNames[generic]} -> ${generic.toLowerCase()}(this.${generic.toLowerCase()})")
+                        }
+                        endControlFlow()
+                    }
+                    .build()
+    )
+}
+
+private fun List<String>.toTypeParameters() = map { TypeVariableName(it) }
+
+private fun parameterizedCoproductNClassName(generics: List<String>): ParameterizedTypeName =
+        ClassName("", "Coproduct${generics.size}")
```

Comment:
I find the indentation excessive. Not sure if it's your IDE or detekt configuration.